Allow saving / loading from Huggingface Hub preset

Wauplin opened the pull request:
Solves https://github.com/keras-team/keras-nlp/issues/1294. As mentioned in https://github.com/keras-team/keras-nlp/issues/1294#issuecomment-1966864503, this PR adds support for the `hf://` prefix to load presets from the Huggingface Hub. 

The integration requires the [`huggingface_hub`](https://github.com/huggingface/huggingface_hub) library. Authentication can be configured with the `HF_TOKEN` environment variable (only for private models or for uploads, similarly to KaggleHub).  Here is a [Colab notebook](https://colab.research.google.com/drive/1lgA_JHHV_18LNS-KnSmr6f-OkXhPUcB3?usp=sharing) showcasing it.

```py
import keras_nlp
from keras_nlp.models import BertClassifier
from keras_nlp.utils.preset_utils import save_to_preset

classifier = BertClassifier.from_preset("bert_base_en_uncased")
(...) # train/retrain/fine-tune

# Save to Hugging Face Hub
save_to_preset(classifier, "hf://Wauplin/bert_base_en_uncased_retrained")

# Reload from Hugging Face Hub
classifier_reloaded = BertClassifier.from_preset("hf://Wauplin/bert_base_en_uncased_retrained")
```

Here is how it looks like once uploaded on the Hub: https://huggingface.co/Wauplin/bert_base_en_uncased_retrained/tree/main..
If we go this way, I think we should also upload a default model card with `keras-nlp` tag to make all KerasNLP models discoverable on the Hub. On the Hugging Face side, we could make `KerasNLP` an official library (e.g. searchable, with code snippets, download counts, etc.).

In the current implementation, saving to `"hf://Wauplin/bert_base_en_uncased_retrained"` will save the model locally to `Wauplin/bert_base_en_uncased_retrained` subfolder + create the repository on the Hub + upload the local folder to this repo on the Hub. An alternative could be to save to a temporary folder before uploading to the Hub (to avoid the local copy). Both solutions are correct in my opinion, it's more a matter of how the KerasNLP envision the `save_to_preset` method.

mattdangerw commented on keras_nlp/utils/preset_utils.py:
```diff
@@ -109,6 +123,9 @@ def save_to_preset(
     weights_filename="model.weights.h5",
 ):
     """Save a KerasNLP layer to a preset directory."""
+    push_to_hf = preset.startswith(HF_PREFIX)
```
Here we might end up doing some tweaks. We want to support a split flow for saving preprocessing and model weights. So our final flow might need to allow something like this

```python
tokenizer.save_to_preset(dir)
backbone.save_to_preset(dir)
upload_preset("hf://user/model", dir)
```

I don't think we need to solve that here though! @SamanehSaadat is working on a draft of our upload flow currently.

Wauplin replied:
Makes perfect sense!

mattdangerw reviewed (commented):
Looks good! Love how simple this is to add. Let's hold tight for a bit while we nail down the public version of `save_to_preset`.

mattdangerw commented:
@Wauplin with https://github.com/keras-team/keras-nlp/pull/1512 I think we are all ready to go here.

Note that we will keep landing features related to the whole upload flow (in particular uploading a classifier with extra head weights, uploading lora weights that are essentially a diff on another preset). But we will keep the `upload_preset` flow, so I think this is unblocked.

mattdangerw commented on keras_nlp/utils/preset_utils.py:
```diff
@@ -69,6 +76,14 @@ def get_file(preset, path):
+        return huggingface_hub.hf_hub_download(repo_id=hf_handle, filename=path)
     elif os.path.exists(preset):
         # Assume a local filepath.
         return os.path.join(preset, path)
```
Right below this line there is a error message we might want to extend.

Wauplin replied:
Thanks for the hint :+1: Updated it in https://github.com/keras-team/keras-nlp/pull/1510/commits/24ef2626032c78d5c2a02785e557e796f6599160.

mattdangerw reviewed (commented):
Looks great! Just some comments or error messages.

SamanehSaadat commented:
Thanks for the PR, @Wauplin! Looks great! 
Regarding documentation, for now, you can update the `upload_preset()` docstring [here](https://github.com/keras-team/keras-nlp/blob/978542359a16a51778cb0b9434f1cdb2fdba677b/keras_nlp/utils/preset_utils.py#L260-L263) to include info about `hf://`.
I'm working on preparing a Kaggle upload guide and I'll make sure to include info about HuggingFace as well.

Wauplin commented on keras_nlp/utils/preset_utils.py:
```diff
@@ -69,6 +76,14 @@ def get_file(preset, path):
+                "Please install with `pip install huggingface_hub`."
+            )
+        hf_handle = preset.removeprefix(HF_PREFIX)
+        return huggingface_hub.hf_hub_download(repo_id=hf_handle, filename=path)
```
At this point, `hf_handle` must correspond to a repo_id so something in the form `username/repo_name` (e.g. "google/gemma-7b") since the hf prefix has been removed. If it's not the case, an `HFValidationError` (which is a custom `ValueError`) is raised. [Here are the validation rules we are checking.](https://github.com/huggingface/huggingface_hub/blob/a9453d9f08b1d8ec926f57cc3f18f5902021b7cc/src/huggingface_hub/utils/_validators.py#L124)

Wauplin replied:
I just pushed https://github.com/keras-team/keras-nlp/pull/1510/commits/24ef2626032c78d5c2a02785e557e796f6599160 to raise a ValueError similar to the kaggle handle one. I tried to be as consistent as possible. Please let me know what you think :)

Wauplin commented on keras_nlp/utils/preset_utils.py:
```diff
@@ -262,6 +277,15 @@ def upload_preset(
+        huggingface_hub.upload_folder(repo_id=repo_url.repo_id, folder_path=preset)
     else:
         raise ValueError(
             f"Unexpected URI `'{uri}'`. Kaggle upload format should follow "
```
I've update the error message in https://github.com/keras-team/keras-nlp/pull/1510/commits/24ef2626032c78d5c2a02785e557e796f6599160. I kept the `uri` naming to be consistent with the rest of the logic but agree with you it would be good to harmonize naming. I think inconsistency comes from the fact that in `get_file` the `preset` refers to either a local directory or a URI while in `upload_preset`, the `preset` refers to a local directory and `uri` to the URI. So maybe not so inconsistent after all?

Wauplin commented:
Thanks both for the review and feedback! I have addressed all comments and completed the `upload_preset` docstring to mention `hf://`. I did not add it to `get_file` since kaggle preset was not documented there either but happy to document them both if you think it makes sense.

Wauplin commented on keras_nlp/utils/preset_utils.py:
```diff
@@ -262,10 +288,29 @@ def upload_preset(
-            f"Unexpected URI `'{uri}'`. Kaggle upload format should follow "
-            "`kaggle://<KAGGLE_USERNAME>/<MODEL>/<FRAMEWORK>/<VARIATION>`."
+            "Unknown URI. An URI must be a one of:\n"
+            "1) a Kaggle Model handle like `'kaggle://<KAGGLE_USERNAME>/<MODEL>/<FRAMEWORK>/<VARIATION>'`\n"
```
Here I followed the existing message but I found it inconsistent with the error in `get_file`. In `get_file`, we provide real examples (e.g. `'kaggle://keras/bert/keras/bert_base_en'`) while here we only provide the format (`'kaggle://<KAGGLE_USERNAME>/<MODEL>/<FRAMEWORK>/<VARIATION>'`). Both are fine I think but if you prefer one or the other, please let me know and I can update in this PR.

mattdangerw replied:
I think we can merge this as is, and I'll chat with folk later to figuring out our broader naming a push a small fix.

mattdangerw reviewed (approved):
Thanks so much! Looking forward to being on the hub!

Wauplin commented:
Great thanks for the approval! I just pushed a commit to fix linting. Hope it's fine now :)

SamanehSaadat reviewed (approved):
Thanks a lot!

SamanehSaadat commented:
Hi @Wauplin!

Thanks again for the PR!
I was trying to create a demo for HF upload and realized when you want to upload a model, you need to create the model on the HF web UI first.  `upload_folder()` [documentation](https://huggingface.co/docs/huggingface_hub/guides/upload#upload-a-folder) mentions that the folder can be uploaded to an **existing repo**.

Just wanted to make sure my understanding is correct and there isn't any way to upload a model folder if the model hasn't been created on the HF website.

Wauplin commented:
> I was trying to create a demo for HF upload and realized when you want to upload a model, you need to create the model on the HF web UI first. upload_folder() [documentation](https://huggingface.co/docs/huggingface_hub/guides/upload#upload-a-folder) mentions that the folder can be uploaded to an existing repo.

@SamanehSaadat yes that's true but you can use `create_repo` to create the repo on the Hub first. This is actually what we are doing [here](https://github.com/keras-team/keras-nlp/pull/1510/files#diff-c0945af0d5354fabef85ad6ce282e20b0eb2fd4942113ee42e4bd2dd4331d84fR301).
